server/session_jwe.go

```diff
@@ -0,0 +1,75 @@
+package server
+
+import (
+	"crypto/rsa"
+	"fmt"
+
+	"github.com/crewjam/saml"
+	"github.com/crewjam/saml/samlsp"
+	"github.com/go-jose/go-jose/v3"
+)
+
+// JWESessionCodec wraps a standard SessionCodec and applies JWE encryption to protect sensitive attributes
+type JWESessionCodec struct {
+	wrapped    samlsp.SessionCodec
+	encrypter  jose.Encrypter
+	privateKey *rsa.PrivateKey
+}
+
+func NewJWESessionCodec(wrapped samlsp.SessionCodec) (samlsp.SessionCodec, error) {
+	// get the public and private key from the underlying codec to use for encryption
+	codec, ok := wrapped.(samlsp.JWTSessionCodec)
+	if !ok {
+		return nil, fmt.Errorf("wrapped session codec is not a JWTSessionCodec")
+	}
+	publicKey := &codec.Key.PublicKey
+	privateKey := codec.Key
+
+	// create a JWE encrypter (possible to parameterize jose.ContentEncryption and jose.KeyAlgorithm)
+	encrypter, err := jose.NewEncrypter(jose.A128GCM, jose.Recipient{Algorithm: jose.RSA_OAEP, Key: publicKey}, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create jwe encrypter: %w", err)
+	}
+
+	return JWESessionCodec{wrapped: wrapped, encrypter: encrypter, privateKey: privateKey}, nil
+}
+
+func (c JWESessionCodec) New(assertion *saml.Assertion) (samlsp.Session, error) {
```

Comment:
I wonder if these should be a `*JWESessionCodec` however looks like the struct is just all pointers

Reply:
good catch